Makefile: speed up `buildshort` and introduce `cleanshort` target

irfansharif opened the pull request:
Makefile: speed up `buildshort` and introduce `cleanshort` target

I took some time to investigate where all time was going every \`make
buildshort\` invocation, and here are a few changes to tighten it up a
bit. Here's what my workflow tends to look like:

```sh
$ make clean       # Used gratuitously, like `clear`. Probably shouldn't.
$ make buildshort  # To check if everything compiles.
$ make buildshort  # As I'm iterating on things, to keep checking things.
```

Almost all of my changes are exclusively go code, and as far as
generated code goes, it's only ever protobuf generated code. Others may
also use `make generate` to regenerate code (including execgen, optgen,
etc.), but I assume the workflow is roughly the same for all. I also
assume protobufs are changed far more often than `execgen` or `optgen`
code.

To speed things up for most folks, this PR aims to make the common
sequences of "iteratively compile crdb" and "clean up some stuff and
compile crdb" faster than it was before. It does so by doing the
following:

1. Not tagging `build.utcTime` on every `buildshort` invocation, which
   lets us make much better use of the go build cache.

```sh
$ make clean; time make buildshort       # Fresh builds before
  156.40 real       553.84 user       380.26 sys
$ make clean; time make buildshort       # Fresh builds after
  121.04 real       384.52 user       278.03 sys

$ make buildshort; time make buildshort  # No-op builds before
  27.44 real        21.91 user        32.50 sys
$ make buildshort; time make buildshort  # No-op builds after
  4.79 real         3.21 user        30.79 sys
```

2. Not rebuilding the settings HTML page or the docgen page for
   `buildshort`. Given we're not building the UI, I think this is fine
   (and shaves a few seconds)

3. Introducing a `cleanshort` target, which leaves the go build cache
and C++ artifacts intact. It should suffice for most folks editing
`.proto` files and regular go code. `{exec,opt}gen` and storage folks
may still need to fall back to `make {clean,generate}`, but nobody else.

```sh
$ make clean; time make buildshort        # Fresh build; nuking most things.
  121.04 real       384.52 user       278.03 sys

$ make cleanshort; time make buildshort   # Fresh-ish build; nuking only go protos.
  31.10 real        29.04 user        36.76 sys
```

---

We also add a loud warning if `ccache` is not present on the
maintainer's system, as it really should be.

```sh
$ make clean; time make build   # Without ccache
409.45 real      3251.96 user       598.34 sys

$ make clean; time make build   # With ccache
156.40 real       553.84 user       380.26 sys
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45611)
<!-- Reviewable:end -->

petermattis commented:
> Try our best not to rebuild execgen generated code.

Is rebuilding the execgen code problematic, or the fact that we're updating the timestamps even when the source hasn't changed? I can help if the problem is the latter.

irfansharif commented:
Both, but more than latter. But I hadn't thought of that approach, I was
toying around with something like `cmp --silent <old> <new> || mv <new>
<old>` for other things, and perhaps it also applies here. I'll try it
tomorrow.

On Mon, Mar 2, 2020, 8:04 PM Peter Mattis <notifications@github.com> wrote:

> Try our best not to rebuild execgen generated code.
>
> Is rebuilding the execgen code problematic, or the fact that we're
> updating the timestamps even when the source hasn't changed? I can help if
> the problem is the latter.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/45611?email_source=notifications&email_token=ACQMN4V6WJBL6A4L77GO6GDRFRJQDA5CNFSM4K773V7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENRVI5Q#issuecomment-593712246>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQMN4QH6NGJRXTOA5HETJDRFRJQDANCNFSM4K773V7A>
> .
>

irfansharif commented:
Hm, `execgen` doesn't really work with unix IO very well it seems. It just
takes in specific file paths and overwrites them. Short of changing that
(which we probably should anyway), did you have something else in mind to
not modify timestamps?

On Mon, Mar 2, 2020 at 9:04 PM irfan sharif <notifications@github.com>
wrote:

> Both, but more than latter. But I hadn't thought of that approach, I was
> toying around with something like `cmp --silent <old> <new> || mv <new>
> <old>` for other things, and perhaps it also applies here. I'll try it
> tomorrow.
>
> On Mon, Mar 2, 2020, 8:04 PM Peter Mattis <notifications@github.com>
> wrote:
>
> > Try our best not to rebuild execgen generated code.
> >
> > Is rebuilding the execgen code problematic, or the fact that we're
> > updating the timestamps even when the source hasn't changed? I can help
> if
> > the problem is the latter.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/cockroachdb/cockroach/pull/45611?email_source=notifications&email_token=ACQMN4V6WJBL6A4L77GO6GDRFRJQDA5CNFSM4K773V7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENRVI5Q#issuecomment-593712246
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ACQMN4QH6NGJRXTOA5HETJDRFRJQDANCNFSM4K773V7A
> >
> > .
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/45611?email_source=notifications&email_token=ACQMN4XYINCLBTR6IPJTBMTRFRQQ5A5CNFSM4K773V7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENRZAHA#issuecomment-593727516>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQMN4VEQL5JJBNSBONI3BDRFRQQ5ANCNFSM4K773V7A>
> .
>


-- 
Irfan Sharif, Staff Of Technical Member
*Cockroach Labs*

jordanlewis commented:
I'm down with all of this cool stuff - but I never, ever run `make clean`! Am I an outlier? Why do you run `make clean`?

If you have a suggestion for how to make `execgen` better behaved let me know, I think I just copied how `optgen` works so we'll have to fix both.

irfansharif commented:
> Why do you run make clean?

I don't have a good reason, but:
- sometimes if I cancel a `make` invocation midway through code generation, I can have a codebase that will not compile (failing to import "cockroachdb/cockroach/pkg/fmt" and the like, because we `sed` replace them after generating them, for whatever reason). 
- other times I run it when switching to old commits and finding extraneous *.eg.go files (I know we added the right `.gitignore` listings in https://github.com/cockroachdb/cockroach/pull/35858 so this isn't a problem anymore, but I've jumped to commits where that wasn't there).
- if `make {build,generate,buildshort,stress,etc}` fails for whatever reason, `make clean` is a low-effort hammer to use.
- just muscle memory, I run it for the same reason I `clear` my terminal windows.

> If you have a suggestion for how to make execgen better behaved let me know, I think I just copied how optgen works so we'll have to fix both.

I was just thinking out loud here, but it'd be nice if it had the following usage:
```
execgen <input>_tmpl.go -o <out>.eg.go
execgen <input>_tmpl.go > <out>.eg.go
```

That way it wouldn't just automatically rewrite the eg.go files, and cause `go build` to redo all the work.

jordanlewis commented:
How would that method of invocation help matters? The annoying thing about dependencies and execgen is that if any transitive dependencies of the `execgen` package change (which includes like, `pkg/sql/sem/tree` and some other ones that change pretty frequently) you technically need to make sure to regenerate all `.eg.go` files - that's as fine-grained as the dependency tracking goes.

petermattis commented:
> How would that method of invocation help matters?

The standard method of generating source is to generate to a temp file and only rename to the actual file if the contents differ. `execgen` still gets rebuilt, but if the generated source doesn't actually change, the rebuilding stops there.

knz commented:
> The standard method of generating source is to generate to a temp file and only rename to the actual file if the contents differ. 

I was about to write the same. In fact I am surprised the Makefile is not already written this way. The standard structure for a make rule is

```
output: input
    gen input >output.tmp || rm -f output.tmp
    mv output.tmp output
```

petermattis commented:
I think that last line should be something like:

```
    cmp output.tmp output && rm -f output.tmp || mv output.tmp output
```

jordanlewis commented:
oh, that's smart! I'll take a stab at it later.

jordanlewis commented:
Wait, isn't Go already going to no-op part of the build if the file is the same? I thought it had a cache...

irfansharif commented:
Go does have a build cache, but it seems it doesn't check the contents of
the files, and instead relies on mtime/ctime (I'm guessing). If you try
regenerating the *.eg.go files and rebuilding, you'll see the `go build`
process log it's attempt to compile pkg/sql/colexec.

On Wed, Mar 4, 2020, 10:15 PM Jordan Lewis <notifications@github.com> wrote:

> Wait, isn't Go already going to no-op part of the build if the file is the
> same? I thought it had a cache...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/45611?email_source=notifications&email_token=ACQMN4U6JV674OV5LMAJ7ATRF4KLXA5CNFSM4K773V7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEN3RJFQ#issuecomment-595006614>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQMN4UE24K25I7W43ZVUWDRF4KLXANCNFSM4K773V7A>
> .
>

petermattis commented:
> Go does have a build cache, but it seems it doesn't check the contents of
the files, and instead relies on mtime/ctime (I'm guessing).

This is true of many build caches. There was a later generation of build tools which used the fingerprints of the input contents to invalidate the cache. Bazel falls into this category. I feel like I'm beating a dead horse, but we should really move away from make. All these frustrations we're having have been fixed by a better tool. Bazel is that better tool (though I won't promise it comes without its own frustrations).

knz commented:
What I learned from Bob who was investigating this, is that Bazel and the Go build system are waging war about what they consider are dependencies. 

We can either enter that battlefield to help Bazel, which will be as painful as figuring the current `make` trouble; or we can let them fight without a clear outcome, in which case Bazel may not deliver as much benefit as it says on the label.

All in all I don't disagree we can use better tools, but we should also not under-estimate the complexity of the task and budget it appropriately. (I personally think this deserves dedicated effort over a couple of weeks, which doesn't fit the interrupt-driven schedule of the current dev-inf team.)

petermattis commented:
@knz I didn't mean to suggest we could switch to Bazel without dedicated effort. Doing so will definitely be a large lift. I'd personally rather do that lift than fight with make. This is speaking from long make experience and long experience with Blaze (the Google interval version of Bazel).

knz commented:
no objection from me - I just wanted to point out that Bazel is not a magic instrument. Someone needs to tell it about the dependency graph. That is no joke. There are hundreds of edges in our Makefile's dep graph already.

irfansharif commented:
> What I learned from Bob who was investigating this, is that Bazel and the Go build system are waging war about what they consider are dependencies.

Do you mind elaborating here? For my own curiosity.

knz commented:
> Do you mind elaborating here? For my own curiosity.

the go compiler draws its own dep graph when it processes your build but does not provide a tool to export it.  So bazel has to figure it out on its own, separately.

It also is unable to reuse already-build artifacts  with cgo and instead always re-compiles the external objects, even if another build system "did the right thing".

Also go has its own cache of intermediate build artifacts (which is not documented) and so bazel has a hard time figuring out what are the reusable intermediate artifacts for the next build (or, potentially, a concurrent build that happens to be dependent on the same version).

There are other things but I did not do the analysis (Bob did) and this is what what I recollect from a casual chat.

petermattis commented:
> the go compiler draws its own dep graph when it processes your build but does not provide a tool to export it. So bazel has to figure it out on its own, separately.

This is true. Go is a bit special in that the compiler knows the dependency graph. Go is also a bit special in that the compiler exports the dependency graph in an easy to use format. It isn't hard for Bazel to figure it out, and then it records it in BUILD files. The gazelle tool provides a mechanism for automatically updating this dependency graph. It is all pretty straightforward to use. 

> It also is unable to reuse already-build artifacts with cgo and instead always re-compiles the external objects, even if another build system "did the right thing".

Bazel can use the output of an external tool. I'm not sure what you mean by it always recompiles. Certainly, Bazel has to call in to the external tool in oder to determine that a recompilation is necessary. I recall there is something you have to specify to terminate the rebuild chain there if the external dependency did not change.

> Also go has its own cache of intermediate build artifacts (which is not documented) and so bazel has a hard time figuring out what are the reusable intermediate artifacts for the next build (or, potentially, a concurrent build that happens to be dependent on the same version).

My understanding is that Bazel takes complete control of the caching of intermediate artifacts. In particular, Bazel does not use the `go` tool, but invokes the `gc` compiler directly. Pretty sure this is what you want to happen.

> There are other things but I did not do the analysis (Bob did) and this is what what I recollect from a casual chat.

Ack on where this is coming from. I feel there is a bit of hearsay.

Interesting nugget 1: the behavior of the `go` tool for builds is essentially what you get from Bazel if you eliminated BUILD files. This was actually the intent of the Go folks: provide the experience of Bazel (Blaze) without BUILD files. Minimalism, as you might expect. This is possible because Go source files have enough information to construct the dependency graph. But as soon as you introduce another language or generated files this goes out the window. All BUILD files are doing is making this dependency information explicit in a readable and scalable manner. Scalable in the sense that you generally do not have a monolithic BUILD file, but one per package. Readable in the sense that groups of files are first class objects.

Interesting nugget 2: all Go inside Google is built using Bazel (Blaze). Ok, I don't know that it is all Go, but I was told that stat at some point and have no reason to believe it has changed. These builds involve a mixture of generate code (protobufs, go_generic) and other languages (C++).

knz commented:
> > It also is unable to reuse already-build artifacts with cgo and instead always re-compiles the external objects, even if another build system "did the right thing".
>
> Bazel can use the output of an external tool. I'm not sure what you mean

you took "It" with the wrong referent in my first sentence. I meant to say "go is unable to reuse already-built artifats with cgo and instead always re-compiles". In particular the "go" command is unable to benefit from proper artifact management by bazel directly

This concern obviously goes away if bazel indeed invokes "gc" directly. I was not aware this was a thing. I hope it will know about fishing the cgo build flags from .go flags through. That one is tricky to get right and even trickier to fit in a BUILD file manually.

petermattis commented:
> you took "It" with the wrong referent in my first sentence. I meant to say "go is unable to reuse already-built artifats with cgo and instead always re-compiles". In particular the "go" command is unable to benefit from proper artifact management by bazel directly

Heh, I reread your sentence again and still read it the same way. Regardless, I don't think this is a problem. Bazel will only rebuild the C++ artifacts when needed. It should terminate the build update if they have not changed.

> This concern obviously goes away if bazel indeed invokes "gc" directly. I was not aware this was a thing. I hope it will know about fishing the cgo build flags from .go flags through. That one is tricky to get right and even trickier to fit in a BUILD file manually.

I don't know off-hand if it does or doesn't, but if I were to hazard a guess, it doesn't. The cgo build flags in .go files is limited, and fragile. The `go` tool doesn't have a configuration file, so they had to encode this configuration in comments. In Bazel you encode the configuration in the BUILD files. The BUILD file allows this information to be more dynamic, changing as you change the compilation target. Right now, we have to rewrite the `_zcgo.go` files, and use build tags to specify which one to use. This works, but it took a lot of blood and sweat to achieve that.

knz commented:
@irfansharif I did improve the execgen situation a while ago, but I think the remainder of your changes here (especially not tickling the time tag in buildshort) is good. Do you want to rebase?

irfansharif committed:
ui,docs: fix README reference to generated filename

And remove a repeated 'the' somewhere.

Release note: None

irfansharif committed:
Makefile: speed up `buildshort` and introduce `cleanshort` target

I took some time to investigate where all time was going every `make
buildshort` invocation, and here are a few changes to tighten it up a
bit. Here's what my workflow tends to look like:

```sh
$ make clean       # Used gratuitously, like `clear`. Probably shouldn't.
$ make buildshort  # To check if everything compiles.
$ make buildshort  # As I'm iterating on things, to keep checking things.
```

Almost all of my changes are exclusively go code, and as far as
generated code goes, it's only ever protobuf generated code. Others may
also use `make generate` to regenerate code (including execgen, optgen,
etc.), but I assume the workflow is roughly the same for all. I also
assume protobufs are changed far more often than `execgen` or `optgen`
code.

To speed things up for most folks, this PR aims to make the common
sequences of "iteratively compile crdb" and "clean up some stuff and
compile crdb" faster than it was before. It does so by doing the
following not tagging `build.utcTime` on every `buildshort` invocation,
which lets us make much better use of the go build cache.

```sh
$ make clean; time make buildshort       # Fresh builds before
  156.40 real       553.84 user       380.26 sys
$ make clean; time make buildshort       # Fresh builds after
  121.04 real       384.52 user       278.03 sys

$ make buildshort; time make buildshort  # No-op builds before
  27.44 real        21.91 user        32.50 sys
$ make buildshort; time make buildshort  # No-op builds after
  4.79 real         3.21 user        30.79 sys
```

We also introduce a `cleanshort` target, which leaves the go build cache
and C++ artifacts intact. It should suffice for most folks editing
`.proto` files and regular go code. `{exec,opt}gen` and storage folks
may still need to fall back to `make {clean,generate}`, but nobody else.

```sh
$ make clean; time make buildshort        # Fresh build; nuking most things.
  121.04 real       384.52 user       278.03 sys

$ make cleanshort; time make buildshort   # Fresh-ish build; nuking only go protos.
  31.10 real        29.04 user        36.76 sys
```

We stopped short of not rebuilding the settings HTML page or the docgen
page for `buildshort` (but there are additional seconds to be shaved off
here).

---

We also add a loud warning if `ccache` is not present on the
maintainer's system, as it really should be.

```sh
$ make clean; time make build   # Without ccache
409.45 real      3251.96 user       598.34 sys

$ make clean; time make build   # With ccache
156.40 real       553.84 user       380.26 sys
```

Release note: None

irfansharif commented:
> Do you want to rebase?

Done.

knz commented on Makefile:
```diff
@@ -990,12 +995,12 @@ $(COCKROACH) $(COCKROACHOSS) $(COCKROACHSHORT) go-install:
 buildshort: $(COCKROACHSHORT)
-build buildoss buildshort: $(if $(is-cross-compile),,$(DOCGEN_TARGETS))
-build buildshort: $(if $(is-cross-compile),,$(SETTINGS_DOC_PAGE))
+build buildoss: $(if $(is-cross-compile),,$(DOCGEN_TARGETS))
```
I'd rather have the docgen/settings target remain generated without having to run `make build`.

The reason is that in a regular developer cycle, these need to be updated regularly. 

It's likely that a developer will 1) not run `make build` *at all* and at the same time 2) make ac hange that impacts the generated docs.

I don't know what a good solution is (yet). Have you ascertained how expensive these two targets are? If they are negligible,we could keep them in `buildshort`.

irfansharif replied:
Hm, I agree with you, though the numbers do leave me a bit sad.

```sh
# Building docgen/settings
$ time make cleanshort > /dev/null; time make buildshort > /dev/null; time make buildshort > /dev/null;
        5.83 real         3.99 user         5.86 sys
       32.77 real        38.59 user        12.41 sys
        5.66 real         7.78 user         6.17 sys

# Not building docgen/settings
$ time make cleanshort > /dev/null; time make buildshort > /dev/null; time make buildshort > /dev/null;
        6.45 real         4.19 user         6.24 sys
       28.03 real        31.62 user         7.84 sys
        2.21 real         3.22 user         4.15 sys
```

irfansharif replied:
I'll defer to you on final yay/nay here, removing the build-time ticking we had before already brings the incremental rebuild time down from 26s, so I'm much happier than I was before.

knz replied:
Let's keep the docgen in then. In the common case (after a rebuild) we're looking at a difference of 3 seconds, which is OK. When we get some bazel in there, that problem will alleviate on its own.

irfansharif commented:
bors r+
